### ISSUE TITLE
Preserve sidebar height when toggling sidebar

### ISSUE THREAD

**Issue by Simek** [facebook/docusaurus#4415]
## 🐛 Bug Report

Refs: #4311

Currently the left sidebar do not preserve the height on some of short pages (with long sidebars?), when toggling between collapsed and expanded states.

This is causing the visual "jump" of footer each time user interacts with sidebar mode.

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

## To Reproduce

1. Navigate to: https://v2.docusaurus.io/docs/markdown-features
1. Toggle sidebar
1. (optional) Expand more sidebar groups to increase "jump" height.

EDIT @slorber: note to self: the content is smaller than the sidebar. Don't use the Markdown features page: as of 2022 it is now taller. A better repro page is https://docusaurus.io/tests/docs

## Expected behavior

Footer is not "jumping", sidebar height is preserved.

## Actual Behavior

Footer is "jumping", sidebar height is not preserved.

#### Preview

![jump](https://user-images.githubusercontent.com/719641/110965087-29fefb80-8354-11eb-80ef-c925a954d0e3.gif)

## Your Environment

N/A

## Reproducible Demo

* https://github.com/facebook/docusaurus/tree/master/website

**Comment by tapanchudasama** [comment]
I would like to pick this up, if anybody has not started any work on this.

**Comment by slorber** [comment]
Thanks, nobody is on it, just submit a PR :)

**Comment by Josh-Cena** [comment]
@lex111 What do you think about this? Would a UX like this make more sense?

![Test](https://user-images.githubusercontent.com/55398995/162602646-c55b348c-4864-408d-a64a-f6511cd69d04.gif)

I don't know a very robust way to implement it though, and the more I look at it, the more I think a collapse button with persisted vertical position makes sense (see #6717)

**Comment by lex111** [comment]
It's natural to expect that expanding the content part will shift of elements below, and vice versa. Therefore I don't see why that's such a big issue.

**Comment by yzhe819** [comment]
> **Warning**
> I noticed a sidebar related problem but not sure if this is a new issue. 

The displayed portion of the content changes visually when switching between the collapsed and expanded states of the sidebar. When I collapse the sidebar, the page display jumps from the bottom footer to the content section.

### To Reproduce
1. Navigate to: https://docusaurus.io/docs/deployment
2. Scroll to the bottom
3. Collapse the sidebar
4. Expand the sidebar
5. Try 3 and 4 multiple times (You can notice the vertical shift of web page content)


### Demo

https://user-images.githubusercontent.com/68207314/183418074-48a76574-e194-4b02-b7e0-e2dbf93c7d29.mp4

Before the sidebar is collapsed, the page is at the bottom, but after the sidebar is collapsed and expanded, the page will jump to the `Deploying to Layer0` section.

cc @Josh-Cena

**Comment by Josh-Cena** [comment]
No—I'd say that's not expected! I can reproduce this. Could you open a new bug report? Would be great if you can investigate yourself as well.

**Comment by Josh-Cena** [comment]
The jump happens on every page—the longer the page is, the more significant it gets. Can be easily observed here as well: https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-content-docs

**Comment by yzhe819** [comment]
Cool, I'll open a new issue first

**Comment by slorber** [comment]
What @Simek and @yzhe819 report are the exact same bug.

This behavior is because we toggle from `max-width: 1320px` to `max-width: 1440px`

Changing the dimension of the content container leads to some lines wrapping, leading to the layout shift. It's not the sidebar that changes height, but rather the scroll container that changes dimensions as we toggle the sidebar.

![CleanShot 2022-08-09 at 16 59 46@2x](https://user-images.githubusercontent.com/749374/183685275-6748149e-24df-48a7-814a-bf3be93d9401.png)

![CleanShot 2022-08-09 at 17 00 12@2x](https://user-images.githubusercontent.com/749374/183685359-89bd1eb3-8da6-475a-bdb0-ed240706ee50.png)


You'll notice that the layout shift is more important on larger docs, and at the bottom: this is just because there are more lines wrapping above :)

---

In the end => larger container VS no layout shift: you can only pick one of those. 

I'd say the layout shift is a minor annoyance and usually we want to expand a bit the content container when more space is available.

Technically we could keep the same container size for large desktop screens, but for mobile we absolutely want to increase that container width anyway.

<img width="1002" alt="CleanShot 2022-08-09 at 17 02 14" src="https://user-images.githubusercontent.com/749374/183685988-a1199a46-8691-4092-8d38-e57976f543de.png">

It wouldn't make sense to collapse the sidebar and still keep a small content container (< 500px in the screenshot above).

What I mean is: layout shifts are expected to happen, on purpose, because we want to expand the length of the lines

---

Now, a possible solution to mitigate this layout shift, particularly on very large documents where the layout shift can be significant (like the [bottom of our deployment page](https://docusaurus.io/docs/deployment#deploying-to-cloudflare-pages))

- take some content node in the middle of the screen
- ensure it keeps the exact same viewport position after we have resized the container

Note we already do something similar when switching tabs (cf https://github.com/facebook/docusaurus/pull/5618), that can also produce layout shifts, and have some util already:

```js
const {blockElementScrollPositionUntilNextRender} = useScrollPositionBlocker();
```

Now, this won't prevent new line wraps in the visible viewport. If you want 0 layout shift, I'd say it's your responsibility to ensure the content container keeps the exact same size. You can do that with CSS. I don't think we'd want Docusaurus to do that by default.

**Comment by Josh-Cena** [comment]
Reverted the title change to prevent confusion with #7918, which are two distinct issues (this one is about maintaining sidebar height; that one is about maintaining page scroll position)

**Comment by Josh-Cena** [comment]
Here's a reproduction on a "simulated large screen" (via zoom level) in case that helps with understanding:

![Test](https://user-images.githubusercontent.com/55398995/183705146-b30d3d5c-a9f9-4b96-93a8-b9d2bd929f53.gif)

Note there's no content layout shift. Only the footer shifts.

**Comment by slorber** [comment]
This is a much better test case, no need to set any zoom level, just click the sidebar collapse button: https://docusaurus.io/tests/docs

**Comment by 0916dhkim** [comment]
I opened a small PR to solve this issue here #8195

**Comment by slorber** [comment]
Note this page looks also weird on a vertical screen, we should fix that too:

https://docusaurus.io/tests/docs

![CleanShot 2022-10-13 at 11 52 52](https://user-images.githubusercontent.com/749374/195565394-1574e65e-dcc4-4a3e-851e-826e1452b03d.png)

**Comment by pea-nut-z** [comment]
A reproduction of (https://github.com/facebook/docusaurus/issues/4415#issuecomment-1209597904)


![menu_longer_than_content](https://user-images.githubusercontent.com/66141752/195693853-da5dfaa1-78cf-4dfe-8a8c-f274231ac7a9.gif)


I believe this only happens when the right content is shorter than the navbar menu. The footer shifts because the additional space needed to fit in an open navbar menu longer than the right content. All the right content is longer than the navbar menu now, so this shouldn't be a problem.


Here is an example where the content is longer than the menu and works just fine. However, it is also working because there is no word wrap issue, as mentioned in (https://github.com/facebook/docusaurus/issues/7918)
![content_no_shift](https://user-images.githubusercontent.com/66141752/195693071-f0c89590-3a65-4172-9568-e25eb6b584f5.gif)

The cause for the footer shift should be only the word wrap now. Let me know what you think. Thanks!

**Comment by pea-nut-z** [comment]
> Note this page looks also weird on a vertical screen, we should fix that too:
> 
> https://docusaurus.io/tests/docs
> 
> ![CleanShot 2022-10-13 at 11 52 52](https://user-images.githubusercontent.com/749374/195565394-1574e65e-dcc4-4a3e-851e-826e1452b03d.png)
> 
> There's even a little layout shift on hydration:
> 
> ![CleanShot 2022-10-13 at 11 54 20](https://user-images.githubusercontent.com/749374/195565682-7329d2b2-97e3-4337-810f-815946362dcb.gif) [ ![CleanShot 2022-10-13 at 11 54 20](https://user-images.githubusercontent.com/749374/195565682-7329d2b2-97e3-4337-810f-815946362dcb.gif) ](https://user-images.githubusercontent.com/749374/195565682-7329d2b2-97e3-4337-810f-815946362dcb.gif) [ ](https://user-images.githubusercontent.com/749374/195565682-7329d2b2-97e3-4337-810f-815946362dcb.gif)

Would you like a CSS detection for the vertical screen and then utilize the same navbar reveal on mobile devices - a button in the top left corner?

**Comment by 0916dhkim** [comment]
> Would you like a CSS detection for the vertical screen

So something like `@media (min-aspect-ratio: 8/5)` ?

**Comment by pea-nut-z** [comment]
> > Would you like a CSS detection for the vertical screen
> 
> So something like `@media (min-aspect-ratio: 8/5)` ?

Found this (https://stackoverflow.com/a/5735636/14173622)

**Comment by slorber** [comment]
This is not a problem of vertical/horizontal/ratio but just having more vh than the sidebar + main content

<img width="1914" alt="CleanShot 2022-10-14 at 14 22 46@2x" src="https://user-images.githubusercontent.com/749374/195846004-32e5762c-8a98-4fc2-91be-35bbf0629467.png">


There are good existing repros to both problems here:
- https://docusaurus.io/tests/docs
- https://docusaurus.io/tests/docs/doc-with-another-sidebar

**Comment by pea-nut-z** [comment]
> This is not a problem of vertical/horizontal/ratio but just having more vh than the sidebar + main content
> 
Thanks for explaining that. Here is what I did.
<img width="674" alt="min_height-navbar" src="https://user-images.githubusercontent.com/66141752/195879591-9907b15f-851e-4ea5-a410-2d91290e6699.png">

Move the navbar to its outer container - "main-wrapper," so it will always take up the white space available, giving it a minimum height instead of depending on the content length. 

The result 
![give_nav_min_height](https://user-images.githubusercontent.com/66141752/195880396-af7b7861-4a60-41bd-b9bf-4dc4b041732b.gif)

Is this what you're looking for?

**Comment by slorber** [comment]
@pea-nut-z honestly not sure what I want exactly, need to see it live and play with it to feel the UX on different edge cases.

Can you open a PR with this suggestion?

**Comment by lex111** [comment]
> This is not a problem of vertical/horizontal/ratio but just having more vh than the sidebar + main content

It shouldnt be that way, Ive fixed it erlier, apparently it's broken again https://github.com/facebook/docusaurus/pull/7025

**Comment by 0916dhkim** [comment]
@slorber Would you look at #8328 and let me know what you think?

**Comment by slorber** [comment]
This issues' bug is fixed by @0916dhkim in #8328 👍 thanks

Opened 2 other related docs layout issues discovered in the process:
- https://github.com/facebook/docusaurus/issues/8369
- https://github.com/facebook/docusaurus/issues/8370

**Comment by cainmagi** [comment]
### Current version

Docusaurus 2.3.1

### Description

This issue appears again because someone made a change from:

```css
.main-wrapper {
  display: block;
}
.main-wrapper docsWrapper_XXXX {
  display: flex;
}
```

to 

```css
.main-wrapper {
  display: flex;
  flex-direction: column;
}
.main-wrapper docsWrapper_XXXX {
  display: flex;
  flex-direction: column;
}
```

### How to reproduce

Just initialize the standard template:

```bash
npx create-docusaurus@2.3.1 test-doc classic
cd test-doc
npm start
```

Then go to this page:
```addr
http://localhost:3000/docs/category/tutorial---extras
```

![image](https://user-images.githubusercontent.com/32186723/226462451-bac5e308-4508-47db-9ad4-3bd4bc6ad757.png)

### My solution

I made a hot fixture for this issue by modifying the `src/css/custom.css`:

```css
// Fix a bug caused by Docusaurus, which makes the sidebar not expand again!
.docs-wrapper div.main-wrapper {
  flex-direction: row;
}
```

But I believe that this issue should be solved by modifying the source codes of Docsauraus. Please take a look at the source codes, sir @slorber. Thank you!
